### PR TITLE
Use new IntervalMergingIterator to merge intervals on the fly 

### LINKED ISSUE #1454: Memory Issues with VcfToIntervalList
### Affected tool(s)
VcfToIntervalList

### Affected version(s)
- [x] Latest public release version [version?]
- [x] Latest development/master branch as of [date of test?]

### Description 
Need a ridiculous amount of memory to run VcfToIntervalList because it's holding all of the intervals in memory.

#### Steps to reproduce
Run `VcfToIntervalList` on a large vcf (ex: 5G gzipped vcf), requires >30G of memory
`java -jar picard.jar VcfToIntervalList I=${large_vcf} O=${smaller}.interval_list`

#### Expected behavior
Shouldn't use this much memory 

#### Actual behavior
Uses a ton of memory because it's storing all the intervals in memory. Yossi says "You need to do the sorting and merging in the CLP itself using a disk-backed sorting collection and then merge the intervals as you write them... avoid using the IntervalList object"

### PR DESCRIPTION
**Pull request by yfarjoun**
and write the intervalList without having to hold it all in memory.

fixes #1454 

### Description

_Give your PR a **concise** yet **descriptive** title_
_Please explain the changes you made here._
_Explain the **motivation** for making this change. What existing problem does the pull request solve?_
_Mention any issues fixed, addressed or otherwise related to this pull request, including issue numbers or hard links for issues in other repos._
_You can delete these instructions once you have written your PR description._

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [ ] Added or modified tests to cover changes and any new functionality
- [ ] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

### CODE REVIEW

File: src/main/java/picard/vcf/VcfToIntervalList.java [hunk 1]
```diff
@@ -84,10 +91,23 @@ protected int doWork() {
         IOUtil.assertFileIsReadable(INPUT);
         IOUtil.assertFileIsWritable(OUTPUT);
 
-        final IntervalList intervalList = VCFFileReader.fromVcf(INPUT, INCLUDE_FILTERED);
+        try (VCFFileReader vcfReader = new VCFFileReader(INPUT.toPath(), false)) {
+            final Iterator<Interval> samFileIterator = VCFFileReader.toIntervals(vcfReader, INCLUDE_FILTERED);
+            try (IntervalListWriter writer = new IntervalListWriter(OUTPUT.toPath(), new SAMFileHeader(vcfReader.getFileHeader().getSequenceDictionary()))) {
+                final IntervalList.IntervalMergerIterator mergingIterator =
+                        new IntervalList.IntervalMergerIterator(samFileIterator, true, false, true);
```
Comment (kachulis) [inline src/main/java/picard/vcf/VcfToIntervalList.java]: Would be nice to expose to the command line the option to not concatenate all the merged interval names.  In the use case which prompted this, a TON of sites (think whole genome gvcf with bp resolution) are often merged into a single interval, which then has an outrageously long concatenated name.  These concatenated names often don't carry any useful information, and can make the resulting interval list extremely large.

Comment (skwalker) [reply src/main/java/picard/vcf/VcfToIntervalList.java]: Yeah, 👍

Comment (yfarjoun) [reply src/main/java/picard/vcf/VcfToIntervalList.java]: like so?

Comment (kachulis) [reply src/main/java/picard/vcf/VcfToIntervalList.java]: yep!

Comment (yfarjoun) [reply src/main/java/picard/vcf/VcfToIntervalList.java]: @skwalker can you try this on your VCF that was causing all the trouble and see if it helps?

Comment (skwalker) [reply src/main/java/picard/vcf/VcfToIntervalList.java]: It works! Thank you (the output interval list went from almost 10G to less than 1 with that change)

File: src/main/java/picard/vcf/VcfToIntervalList.java [hunk 2]
```diff
@@ -84,10 +96,23 @@ protected int doWork() {
...
+        try (VCFFileReader vcfReader = new VCFFileReader(INPUT.toPath(), false)) {
+            final Iterator<Interval> samFileIterator = VCFFileReader.toIntervals(vcfReader, INCLUDE_FILTERED);
+            try (IntervalListWriter writer = new IntervalListWriter(OUTPUT.toPath(), new SAMFileHeader(vcfReader.getFileHeader().getSequenceDictionary()))) {
+                final IntervalList.IntervalMergerIterator mergingIterator =
+                        new IntervalList.IntervalMergerIterator(samFileIterator, true, false, CONCATENATE_IDS);
+                for (final Interval interval : new IterableAdapter<>(mergingIterator)){
+                    writer.write(interval);
+                }
+            } catch (IOException e) {
+                if(!OUTPUT.renameTo(new File(OUTPUT.getAbsolutePath() + "incomplete"))){
```
Comment (skwalker) [inline src/main/java/picard/vcf/VcfToIntervalList.java:108]: does this need to be ".incomplete" ?

File: src/main/java/picard/vcf/VcfToIntervalList.java [hunk 3]
```diff
@@ -74,6 +81,11 @@
     @Argument(shortName = StandardOptionDefinitions.OUTPUT_SHORT_NAME, doc = "The output Picard Interval List.")
     public File OUTPUT;
 
+    @Argument(doc="Controls the naming of the resulting intervals. When true, each resulting interval will be named the concatenation of " +
+            "the variant ID fields (if present), or 'interval-<number>' (if not) with a pipe '|' separator. " +
+            "When false, only the first name will be used.")
+    public boolean CONCATENATE_IDS = true;
```
[the remark this answers is not in the document]
Comment (yfarjoun) [reply src/main/java/picard/vcf/VcfToIntervalList.java:87]: good point.

File: src/main/java/picard/vcf/VcfToIntervalList.java (hunk 2 above)
Comment (yfarjoun) [reply src/main/java/picard/vcf/VcfToIntervalList.java:108]: ah, yes. I'll add the period. thanks.

File: src/main/java/picard/util/IntervalListToBed.java [hunk 4]
```diff
@@ -69,24 +69,37 @@ protected int doWork() {
...
             out.close();
-        }
-        catch (IOException ioe) {
+            return 0;
+        } catch (IOException ioe) {
             throw new RuntimeIOException(ioe);
         }
 
-        return 0;
+
```
Comment (yfarjoun) [inline src/main/java/picard/util/IntervalListToBed.java:102]: unneeded empty new lines.

File: src/main/java/picard/vcf/VcfToIntervalList.java [hunk 5]
```diff
@@ -84,7 +84,7 @@
     @Argument(doc="Controls the naming of the resulting intervals. When true, each resulting interval will be named the concatenation of " +
             "the variant ID fields (if present), or 'interval-<number>' (if not) with a pipe '|' separator. " +
             "When false, only the first name will be used.")
-    public boolean CONCATENATE_IDS = true;
+    public boolean CONCATENATE_IDS = false;
```
Comment (yfarjoun) [inline src/main/java/picard/vcf/VcfToIntervalList.java:87]: Hmmm. The reason that this was default `true` was that that was the behavior prior to making this an option....we didn't want to change the default behavior....why does the default need to change?

File: src/main/java/picard/util/IntervalListToBed.java [hunk 6]
```diff
@@ -23,11 +23,11 @@
  */
 package picard.util;
 
-import htsjdk.samtools.util.CollectionUtil;
-import htsjdk.samtools.util.IOUtil;
-import htsjdk.samtools.util.Interval;
-import htsjdk.samtools.util.IntervalList;
-import htsjdk.samtools.util.RuntimeIOException;
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.util.*;
```
Comment (yfarjoun) [inline src/main/java/picard/util/IntervalListToBed.java:27]: there is a preference for explicit imports...can you ask IntelliJ to use explicit imports?

File: src/main/java/picard/vcf/VcfToIntervalList.java (hunk 5 above)
Comment (skwalker) [reply src/main/java/picard/vcf/VcfToIntervalList.java:87]: I thought boolean arguments are supposed to default to false? I could make it 
 `DONT_ CONCATENATE_IDS` or something or maybe it doesn't matter enough

File: src/main/java/picard/util/IntervalListToBed.java [hunk 7]
```diff
@@ -69,24 +69,37 @@ protected int doWork() {
         IOUtil.assertFileIsReadable(INPUT);
         IOUtil.assertFileIsWritable(OUTPUT);
 
-        IntervalList intervals = IntervalList.fromFile(INPUT);
-        if (SORT) intervals = intervals.sorted();
```
[the remark this answers is not in the document]
Comment (skwalker) [reply src/main/java/picard/util/IntervalListToBed.java:73]: It's causing it to hold everything in place which takes forever to load (I was running a <1 G vcf and it was running for over 3 hours...) and now it takes 3 minutes

File: src/main/java/picard/util/IntervalListToBed.java [hunk 8]
```diff
@@ -69,24 +69,37 @@ protected int doWork() {
...
+        try (final AbstractFeatureReader<Interval, LineIterator> intervalsReader = AbstractFeatureReader.getFeatureReader(INPUT.getPath(), new IntervalListCodec(), false);
+             final BufferedWriter out = IOUtil.openFileForBufferedWriting(OUTPUT)) {
 
-        try {
-            final BufferedWriter out = IOUtil.openFileForBufferedWriting(OUTPUT);
-            for (final Interval i : intervals) {
+            Iterable<Interval> intervalIterable = intervalsReader.iterator();
```
[the remark this answers is not in the document]
Comment (skwalker) [reply src/main/java/picard/util/IntervalListToBed.java:80]: it is in the try clause, I think maybe it looks confusing in the side by side view ?

File: src/main/java/picard/util/IntervalListToBed.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (skwalker) [reply src/main/java/picard/util/IntervalListToBed.java:73]: But maybe I was just using too small a machine in terra? I don't know it actually ran fast on my computer so maybe this is all unnecessary

File: src/main/java/picard/vcf/VcfToIntervalList.java (hunk 5 above)
Comment (yfarjoun) [reply src/main/java/picard/vcf/VcfToIntervalList.java:87]: yeah, make the default value `false`, but keep the default behavior the same...or define an enum "CONCAT_ALL" & "USE_FIRST" to control the behaviour (with the default value being CONCAT_ALL)

Comment (skwalker) [reply src/main/java/picard/vcf/VcfToIntervalList.java:87]: Yeah, I like the enum idea better than a double negative (too confusing) 👍

File: src/main/java/picard/vcf/VcfToIntervalList.java [hunk 9]
```diff
@@ -81,10 +81,17 @@
...
+    public enum VARIANT_ID_TYPES {
+        CONCAT_ALL,
+        USE_FIRST
+    }
+
+    @Argument(doc="Controls the naming of the resulting intervals. When set to CONCAT_ALL (the default), each resulting " +
+            "interval will be named the concatenation of the variant ID fields (if present), or 'interval-<number>' " +
+            "(if not) with a pipe '|' separator. If set to USE_FIRST, only the first name will be used.")
+    public VARIANT_ID_TYPES VARIANT_ID_METHOD = VARIANT_ID_TYPES.CONCAT_ALL;
+    public final boolean concatenate_ids = (VARIANT_ID_METHOD == VARIANT_ID_TYPES.CONCAT_ALL);
```
Comment (yfarjoun) [inline src/main/java/picard/vcf/VcfToIntervalList.java:94]: this cannot work...the @Argument value is modified after the class is instantiated...so `concatenate_ids` will always be true. move this line into `doWork()`

**Comment by skwalker** [comment]
back at you @yfarjoun

File: src/main/java/picard/util/IntervalListToBed.java [hunk 10]
```diff
@@ -69,24 +69,37 @@ protected int doWork() {
...
+            }
+
+            for (final Interval i : intervalIterable) {
                 final String strand = i.isNegativeStrand() ? "-" : "+";
                 final List<?> fields = CollectionUtil.makeList(i.getContig(), i.getStart()-1, i.getEnd(), i.getName(), SCORE, strand);
                 out.append(fields.stream().map(String::valueOf).collect(Collectors.joining("\t")));
                 out.newLine();
-            }
 
+            }
```
[the remark this answers is not in the document]
Comment (yfarjoun) [reply src/main/java/picard/util/IntervalListToBed.java:101]: still no need to close

File: src/main/java/picard/util/IntervalListToBed.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (yfarjoun) [reply src/main/java/picard/util/IntervalListToBed.java:80]: I meant that intervalIterable can be instantiated inside the parens:
```
try (AbstractFeatureReader<Interval, LineIterator> intervalsReader = AbstractFeatureReader.getFeatureReader(INPUT.getPath(), new IntervalListCodec(), false);
       Iterable<Interval> intervalIterable = intervalsReader.iterator();
       BufferedWriter out = IOUtil.openFileForBufferedWriting(OUTPUT)){
```

File: src/test/java/picard/vcf/VcfToIntervalListTest.java [hunk 11]
```diff
@@ -48,12 +49,33 @@ public void testExcludingFiltered(
...
         Assert.assertEquals(intervals.size(), expectedIntervalsSize);
+
+        if (useFirstID) {
+            for (Interval interval : intervals) {
+                Assert.assertFalse(interval.getName().contains("|"));
+            }
+        } else {
+            // make sure the one where two sites that should be concatenated into one interval were actually concatenated
+            Assert.assertTrue(intervals.get(5).getName().contains("|"));
+
```
Comment (yfarjoun) [inline src/test/java/picard/vcf/VcfToIntervalListTest.java:72]: too many new lines

File: src/test/java/picard/vcf/VcfToIntervalListTest.java [hunk 12]
```diff
@@ -48,12 +49,33 @@ public void testExcludingFiltered(
...
+
+        }
+
+
+
     }
 }
+
+// USE_FIRST
```
Comment (yfarjoun) [inline src/test/java/picard/vcf/VcfToIntervalListTest.java:81]: remove comment?

**Comment by skwalker** [comment]
@yfarjoun I _think_ I made the right changes?

File: src/main/java/picard/util/IntervalListToBed.java (hunk 4 above)
Comment (kachulis) [reply src/main/java/picard/util/IntervalListToBed.java:116]: empty new line still here

File: src/main/java/picard/vcf/VcfToIntervalList.java [hunk 13]
```diff
@@ -83,11 +101,25 @@
     protected int doWork() {
         IOUtil.assertFileIsReadable(INPUT);
         IOUtil.assertFileIsWritable(OUTPUT);
+        final boolean concatenate_ids = (VARIANT_ID_METHOD == VARIANT_ID_TYPES.CONCAT_ALL); // TODO: move to doWork() section
```
Comment (kachulis) [inline src/main/java/picard/vcf/VcfToIntervalList.java:105]: no longer needs "TODO" comment

File: src/main/java/picard/vcf/VcfToIntervalList.java [hunk 14]
```diff
@@ -83,11 +101,25 @@
...
+            try (IntervalListWriter writer = new IntervalListWriter(OUTPUT.toPath(), new SAMFileHeader(vcfReader.getFileHeader().getSequenceDictionary()))) {
+                final IntervalList.IntervalMergerIterator mergingIterator =
+                        new IntervalList.IntervalMergerIterator(samFileIterator, true, false, concatenate_ids);
+                for (final Interval interval : new IterableAdapter<>(mergingIterator)){
+                    writer.write(interval);
+                }
+            } catch (IOException e) {
+                if(!OUTPUT.renameTo(new File(OUTPUT.getAbsolutePath() + ".incomplete"))){
+                    OUTPUT.delete();
+                };
```
Comment (kachulis) [inline src/main/java/picard/vcf/VcfToIntervalList.java:117]: unnecessary semicolon

File: src/test/java/picard/vcf/VcfToIntervalListTest.java (hunk 11 above)
Comment (kachulis) [reply src/test/java/picard/vcf/VcfToIntervalListTest.java:72]: still too many new lines

File: src/test/java/picard/util/IntervalListToBedTest.java [hunk 15]
```diff
@@ -13,6 +13,7 @@
 public class IntervalListToBedTest {
     private static final String TEST_DATA_DIR = "testdata/picard/util/";
     private final File INTERVAL_LIST = new File(TEST_DATA_DIR, "interval_list_to_bed_test.interval_list");
+    private final File UNSORTED_INTERVAL_LIST = new File(TEST_DATA_DIR, "unsorted_interval_list_to_bed_test.interval_list");
```
Comment (kachulis) [inline src/test/java/picard/util/IntervalListToBedTest.java:16]: need to add this file, tests fail currently because doesn't exist in repo

Comment (skwalker) [reply src/test/java/picard/util/IntervalListToBedTest.java:16]: 🤦‍♀️

**Comment by yfarjoun** [comment]
🎉

**Pull request merged**